### PR TITLE
Dominick/create and edit group pages

### LINKED ISSUE #21: [Group] Create/Edit Group Pages - (FE1)
**Build pages for creating and editing Secret Santa groups**

**Create Screen:**

Image

**Edit  Screen:**

Image

Required Fields:
- Group name
- Description
- Image selection
- Budget
- Drawing date
- Exchange date

Acceptance Criteria:
- [ ]  Create new groups
- [ ] Edit existing groups
- [ ] Delete groups
- [ ] Basic form validation

### PR DESCRIPTION
**Pull request by HoldUpFjord**
https://github.com/user-attachments/assets/8d87edca-ac7f-4942-b301-656774de3293

Closes #21 
- [X] Create new groups
- Adds create new groups page
- [X] Edit existing groups
- Adds edit groups page
- [X] Delete groups
- Delete button on edit page
- [X] Basic form validation
- Zod form validation for Typesafety

### CODE REVIEW

File: app/create-group-page/page.tsx [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+                    </FormDescription>
+                    <FormMessage />
+                  </FormItem>
+                )}
+              />
+              <div className="flex justify-start m-5">
+                <Button className="m-2" type="submit">
+                  Create Group
+                </Button>
+              </div>
```
Comment (guardiansofthegit-collab) [inline app/create-group-page/page.tsx:318]: Awesome, the BE part of this is done so you should be able to make a post request on `/api/gift-exchanges`.

(this can be in separate PR)

File: app/create-group-page/page.tsx [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+                        </FormControl>
+                      </PopoverTrigger>
+                      <PopoverContent className="w-auto p-0" align="start">
+                        <Calendar
+                          mode="single"
+                          selected={field.value}
+                          onSelect={field.onChange}
+                          // disabled={(date) =>
+                          //   date > new Date() || date < new Date("1900-01-01")
+                          // }
```
Comment (thomasnguyen) [inline app/create-group-page/page.tsx:302]: question: If we wanted to add validation so that the exchange date must always be after the drawing date, what would it take?

A low priority for the hackathon but a small thing to add if it doesn't take too much time.

File: app/create-group-page/page.tsx [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+  { label: "$10 - $20", value: "10-20" },
+  { label: "$20 - $30", value: "20-30" },
+  { label: "$30 - $40", value: "30-40" },
+  { label: "$40 - $50", value: "40-50" },
+  { label: "$50 - $60", value: "50-60" },
+  { label: "$60 - $70", value: "60-70" },
+  { label: "$70 - $80", value: "70-80" },
+  { label: "$80 - $90", value: "80-90" },
+  { label: "$90 - $100", value: "90-100" },
+] as const;
```
Comment (thomasnguyen) [inline app/create-group-page/page.tsx:49]: So the DB will only take in a singular value, should we use the upper bound when sending over the values to the backend?

File: app/edit-group-page/page.tsx [hunk 4]
```diff
@@ -0,0 +1,335 @@
+"use client";
+
+import { cn } from "@/lib/utils";
+import { zodResolver } from "@hookform/resolvers/zod";
+import { useForm } from "react-hook-form";
+import { z } from "zod";
```
Comment (thomasnguyen) [inline app/edit-group-page/page.tsx:6]: Question: Wondering if it we should move this file to `/gift-exchanges/[id]/edit` so that we can make the correct fetch request to get the data based off of the gift-exchange id

**Review by thomasnguyen — APPROVED** [review]
overall looks good to me. Some of the changes can be addressed in a future PR imo

File: app/create-group-page/page.tsx (hunk 2 above)
Comment (HoldUpFjord) [reply app/create-group-page/page.tsx:302]: It might not be dead-simple, but I think it would be something like creating a variable `selectedDrawingDate` for the selecteddate and then assigning that variable into the `disabled` object in the exchange date form. I'll timeblock this and look into it later today.

File: app/create-group-page/page.tsx (hunk 3 above)
Comment (HoldUpFjord) [reply app/create-group-page/page.tsx:49]: I misunderstood how we'd be formatting the data. I like that solution for the sake of the hackathon. I'll spike on how this data is outputted and see if I can include that in this PR

File: app/edit-group-page/page.tsx (hunk 4 above)
Comment (HoldUpFjord) [reply app/edit-group-page/page.tsx:6]: We could do that! Without the logic written to dynamically pull in that unique pages content, it would just show default page as is correct?

File: app/create-group-page/page.tsx (hunk 2 above)
Comment (thomasnguyen) [reply app/create-group-page/page.tsx:302]: gotcha, yeah if its not simple, I say skip it

File: app/edit-group-page/page.tsx (hunk 4 above)
Comment (thomasnguyen) [reply app/edit-group-page/page.tsx:6]: Yeah that sounds good to me!

File: app/create-group-page/page.tsx (hunk 2 above)
Comment (kepsteen) [reply app/create-group-page/page.tsx:302]: @HoldUpFjord If you want to add this validation, it shouldn't be too hard with Zod. You can use the .refine method to create some custom validation that compares the two dates. https://zod.dev/?id=refine

File: app/create-group-page/page.tsx (hunk 3 above)
Comment (thomasnguyen) [reply app/create-group-page/page.tsx:49]: No worries! I could of probably communicated better on that part. As you can tell, a lot of the details were rushed on my end :P

File: app/edit-group-page/page.tsx [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+          type="button"
+          onClick={() => router.push("/dashboard")}
+        >
+          <ChevronLeft className="mr-2" />
+          Back to Dashboard
+        </Button>
+      </div>
+      <div className="flex items-center justify-center h-full">
+        <div className="bg-white w-1/2 mb-5 flex justify-center align-center rounded flex-col ">
+          <h2 className="font-bold m-5">Edit Secret Santa Page</h2>
```
Comment (kepsteen) [inline app/edit-group-page/page.tsx:102]: Should this be "Create Secret Santa Page"?

File: app/create-group-page/page.tsx (hunk 2 above)
Comment (HoldUpFjord) [reply app/create-group-page/page.tsx:302]: @kepsteen Thanks dude I'll look into this

File: app/create-group-page/page.tsx [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+                )}
+              />
+              <FormField
+                control={form.control}
+                name="selectedImage"
+                render={({ field }) => (
+                  <FormItem>
+                    <FormLabel className="m-5">Group Theme Image</FormLabel>
+                    <FormControl>
+                      <ImageSelector {...field} />
```
[the remark this answers is not in the document]
Comment (HoldUpFjord) [reply app/create-group-page/page.tsx:146]: Good catch! I made the choice not to since I thought the current look better aligns with the OG design. Happy to change that if ya'll think it looks better with a consistent width.

File: app/create-group-page/page.tsx [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+                  <FormItem className="flex flex-col">
+                    <FormLabel className="m-5">Price Range</FormLabel>
+                    <Popover>
+                      <PopoverTrigger asChild>
+                        <FormControl>
+                          <Button
+                            variant="outline"
+                            role="combobox"
+                            className={cn(
+                              "w-[200px] justify-between m-5 w-4/5",
```
[the remark this answers is not in the document]
Comment (HoldUpFjord) [reply app/create-group-page/page.tsx:165]: me being dumb lol

Another good catch dude, thanks!

File: app/edit-group-page/page.tsx (hunk 5 above)
Comment (HoldUpFjord) [reply app/edit-group-page/page.tsx:102]: No because there are 2 unique pages, create & edit. The page in question is the Edit page.

Comment (kepsteen) [reply app/edit-group-page/page.tsx:102]: Sorry, I meant to leave this comment in the `create-group-page.tsx`. The header on the create page also reads "Edit Secret Santa Page"

File: app/create-group-page/page.tsx [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+  groupDescription: z.string().min(2, {
+    message: "Group name must be at least 2 characters long",
+  }),
+  groupBudget: z.number().min(0, {
+    message: "Budget must be a positive number",
+  }),
+  giftDrawingDate: z.date(),
+  giftExchangeDate: z.date(),
+  priceRanges: z.string({
+    required_error: "Please select a language.",
```
Comment (alexappleget) [inline app/create-group-page/page.tsx:67]: shouldnt this say `Please select a price range`?

File: app/create-group-page/page.tsx [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+                              ? priceRanges.find(
+                                  (priceRanges) =>
+                                    priceRanges.value === field.value
+                                )?.label
+                              : "Select a price range"}
+                            <ChevronsUpDown className="opacity-50" />
+                          </Button>
+                        </FormControl>
+                      </PopoverTrigger>
+                      <PopoverContent className="w-[200px] p-0">
```
Comment (alexappleget) [inline app/create-group-page/page.tsx:179]: maybe don't use `w-[200px]` it makes tailwind css unresponsive on different screens.

I would instead use `w-52` which is `208px` but `8px` isn't much different. `w-52` is a built in tailwind css that will be able to adapt to screen size rather than just forcing 200px on all screens

File: app/create-group-page/page.tsx [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+                    <FormLabel className="mx-5 mt-5">
+                      Gift Drawing Date
+                    </FormLabel>
+                    <Popover>
+                      <PopoverTrigger asChild>
+                        <FormControl>
+                          <Button
+                            variant={"outline"}
+                            className={cn(
+                              "w-[240px] pl-3 text-left font-normal m-5 w-4/5",
```
Comment (alexappleget) [inline app/create-group-page/page.tsx:236]: same with this `w-[240px]`. use `w-60` which is actually still 240px for tailwind. also you are overlapping `w-4/5` again lol

File: app/create-group-page/page.tsx [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+                    <FormLabel className="mx-5 mt-5">
+                      Gift Exchange Date
+                    </FormLabel>
+                    <Popover>
+                      <PopoverTrigger asChild>
+                        <FormControl>
+                          <Button
+                            variant={"outline"}
+                            className={cn(
+                              "w-[240px] pl-3 text-left font-normal m-5 w-4/5",
```
Comment (alexappleget) [inline app/create-group-page/page.tsx:282]: same here. `w-60` and dont use `w-4/5` or vise versa

**Review by alexappleget — DISMISSED** [review]

File: app/create-group-page/page.tsx (hunk 8 above)
Comment (HoldUpFjord) [reply app/create-group-page/page.tsx:67]: LOL got me. Good catch.

File: app/edit-group-page/page.tsx (hunk 5 above)
Comment (HoldUpFjord) [reply app/edit-group-page/page.tsx:102]: Fixed

File: app/create-group-page/page.tsx (hunk 8 above)
Comment (HoldUpFjord) [reply app/create-group-page/page.tsx]: Fixed 🫡

File: app/create-group-page/page.tsx (hunk 9 above)
Comment (HoldUpFjord) [reply app/create-group-page/page.tsx]: Fixed 🫡

File: app/create-group-page/page.tsx (hunk 8 above)
Comment (HoldUpFjord) [reply app/create-group-page/page.tsx]: FIxed!

**Comment by HoldUpFjord** [comment]
I made the above CSS changes, but in testing the layout the mobile view is... bad. I'd recommend making another issue for mobile responsive layouts for any additional changes.
image

File: app/edit-group-page/page.tsx (hunk 4 above)
Comment (HoldUpFjord) [reply app/edit-group-page/page.tsx]: Fixed 🫡

File: app/create-group-page/page.tsx [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+      giftExchangeDate: new Date(),
+      priceRanges: "",
+    },
+  });
+
+  function onSubmit(values: z.infer<typeof formSchema>) {
+    console.log(values);
+  }
+  return (
+    <div className=" flex justify-center align-center flex-col">
```
Comment (alexappleget) [inline app/create-group-page/page.tsx:89]: The "align-center" could be causing the weird look in mobile view. I just noticed it. I don't believe it exists. Try "items-center"

Comment (alexappleget) [reply app/create-group-page/page.tsx:89]: Or even delete it. This will most likely make the width the full screen and maybe do "xl:w-3/4" for larger screens. Then when the screen gets smaller it'll be 100% width and not squish things

Comment (HoldUpFjord) [reply app/create-group-page/page.tsx:89]: I tried both those changes and the layout didn't update. There's also a lot of overflow on mobile views for the form fields: 
image
 
 I'm happy to fix it. I guess my question to the group is how much of a priority is responsive layouts? I still have another ticket in this sprint and want to make that known to the group
